cmd/hil-vpnd/vpnstates.go

```diff
@@ -0,0 +1,75 @@
+package main
+
+import (
+	"crypto/rand"
+	"errors"
+	"sync"
+)
+
+// Error indicating that we're out of free ports for openvpn to listen on.
+var ErrNoFreePorts = errors.New("There are no free OpenVPN ports")
+
+// A unique identifier for a vpn.
+type UniqueId [128 / 8]byte
+
+// Track the currently existent vpns, available port numbers, etc.
+type VpnStates struct {
+	sync.Mutex
+
+	// The ports used by each vpn.
+	UsedPorts map[UniqueId]uint16
+
+	// A list of free ports, which may be used with new vpns.
+	FreePorts []uint16
+}
+
+// Allocate an empty VpnStates.
+func newStates() *VpnStates {
+	return &VpnStates{
+		UsedPorts: map[UniqueId]uint16{},
+		FreePorts: []uint16{},
+	}
+}
+
+// Allocate a new vpn. Returns a unique id and a port number.
+// May return ErrNoFreePorts if we're out of port numbers to assign.
+func (s *VpnStates) NewVpn() (UniqueId, uint16, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	var id UniqueId
+	if _, err := rand.Read(id[:]); err != nil {
+		return id, 0, err
+	}
+
+	portNo, err := s.allocPort()
+	if err == nil {
+		s.UsedPorts[id] = portNo
+	}
+	return id, portNo, err
+}
+
+func (s *VpnStates) DeleteVpn(id UniqueId) {
+	s.Lock()
+	defer s.Unlock()
+
+	portNo := s.UsedPorts[id]
+	delete(s.UsedPorts, id)
+	s.releasePort(portNo)
+}
+
+// Allocate a new port for a vpn.
+func (s *VpnStates) allocPort() (uint16, error) {
+	if len(s.FreePorts) == 0 {
+		return 0, ErrNoFreePorts
```

Comment:
Just wanted to point out that port 0 is like a wild card port. Although we are sending the error along with it, but in case we forget to check the error when calling it, we may end up having a randomly allocated port. Maybe `return -1`, because that would definitely be an invalid port?

Reply:
The type is uint16, so -1 as such isn't representable. I don't think there's a value we can return that isn't a valid port.

Reply:
in that case let's just leave it as it is.